### PR TITLE
[BEAM-1117] Port Direct runner stateful ParDo to take KeyedWorkItem

### PR DESCRIPTION
**Pull request by kennknowles**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

This is based upon the required PR #1556 but adds the motivating use of key-preservation logic.

### CODE REVIEW

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/EvaluationContextTest.java [hunk 1]
```diff
@@ -104,11 +104,15 @@ public void setup() {
...
 
+    KeyedPValueTrackingVisitor keyedPValueTrackingVisitor = KeyedPValueTrackingVisitor.create();
+    p.traverseTopologically(keyedPValueTrackingVisitor);
+
     BundleFactory bundleFactory = ImmutableListBundleFactory.create();
     graph = DirectGraphs.getGraph(p);
     context =
         EvaluationContext.create(
-            runner.getPipelineOptions(), NanosOffsetClock.create(), bundleFactory, graph);
+            runner.getPipelineOptions(), NanosOffsetClock.create(), bundleFactory, graph,
```
Comment (tgroh) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/EvaluationContextTest.java]: Formatting.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/KeyedPValueTrackingVisitor.java [hunk 2]
```diff
@@ -105,7 +106,12 @@ public void visitValue(PValue value, TransformHierarchy.Node producer) {
   }
 
   private static boolean isKeyPreserving(PTransform<?, ?> transform) {
-    // There are currently no key-preserving transforms; this lays the infrastructure for them
-    return false;
+    // This is a somewhat hacky check for what is considered key-preserving to the direct runner
```
Comment (tgroh) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/KeyedPValueTrackingVisitor.java]: I object to your use of "somewhat"

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoMultiOverrideFactory.java [hunk 3]
```diff
@@ -109,21 +136,59 @@ public StatefulParDo(
...
+                  .and(underlyingParDo.getSideOutputTags().getAll()),
+              input.getWindowingStrategy(),
+              input.isBounded());
 
       return outputs;
     }
   }
+
+  /**
+   * A runner-specific primitive that is just a key-preserving {@link ParDo}, but we do not have the
```
Comment (tgroh) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoMultiOverrideFactory.java]: Haven't you just added that? And/or this looks like a code/documentation mismatch.

Comment (kennknowles) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoMultiOverrideFactory.java]: Done

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/EvaluationContextTest.java (hunk 1 above)
Comment (kennknowles) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/EvaluationContextTest.java]: Done (now autoformat)

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/KeyedPValueTrackingVisitor.java (hunk 2 above)
Comment (kennknowles) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/KeyedPValueTrackingVisitor.java]: Done & verbosified the comment so no one will mistake this for a good idea :-)

**Comment by kennknowles** [comment]
Conflicts, so I'm going to rebase and let the fixups squash in.

**Comment by kennknowles** [comment]
To your broad comment - I hope you are volunteering to implement :-)

Various assumptions and limitations are discovered as we go, but here are things I know need solving (and I will shortly open a PR with the ones I have partly or wholly solved);

 - Setting and deleting timers by ID (I have this roughly "done")
 - Separating user and system timers via namespaced TimerInternals / StepContexts (I have not bothered yet, since ParDo steps and GABW steps are always isolated anyhow)
 - Setting output watermark holds for the timestamps of user timers. Really, all timers, so it could be usefully done globally.
 - Delivering timers to the new `DoFnRunner.onTimer` method in the direct runner (this is easy and done)
 - Debugging why the watermark is at MAX_TIMESTAMP in `@ProcessElement` :-)

I will make the `checkState` fixup shortly.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoMultiOverrideFactory.java [hunk 4]
```diff
@@ -84,16 +90,36 @@ public GbkThenStatefulParDo(ParDo.BoundMulti<KV<K, InputT>, OutputT> underlyingP
     @Override
     public PCollectionTuple expand(PCollection<KV<K, InputT>> input) {
 
-      PCollectionTuple outputs = input
-          .apply("Group by key", GroupByKey.<K, InputT>create())
-          .apply("Stateful ParDo", new StatefulParDo<>(underlyingParDo, input));
+      // A KvCoder is required since this goes through GBK. Further, WindowedValueCoder
+      // is not registered by default, so we explicitly set the relevant coders.
+      KvCoder<K, InputT> kvCoder = (KvCoder<K, InputT>) input.getCoder();
```
Comment (kennknowles) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoMultiOverrideFactory.java]: Done

**Comment by kennknowles** [comment]
retest this please

**Comment by kennknowles** [comment]
retest this please

**Comment by kennknowles** [comment]
Thanks, then. I'll rebase with squashing to prepare for merge, since there are conflicts anyhow.

**Comment by kennknowles** [comment]
retest this please

**Pull request merged**